eventauth/eventauth_test.go

```diff
@@ -234,14 +236,282 @@ func TestAllowedEmptyRoom(t *testing.T) {
 			}
 		}, {
 			"type": "m.room.create",
+			"state_key": "",
 			"sender": "@u1:a",
 			"room_id": "!r1:a",
-			"event_id": "$e2:a",
+			"event_id": "$e3:a",
 			"prev_events": [["$e1", {}]],
 			"content": {"creator": "@u1:a"},
 			"unsigned": {
 				"not_allowed": "Was not the first event in the room"
 			}
+		}, {
+			"type": "m.room.message",
+			"sender": "@u1:a",
+			"room_id": "!r1:a",
+			"event_id": "$e4:a",
+			"content": {"body": "Test"},
+			"unsigned": {
+				"not_allowed": "No create event"
+			}
+		}, {
+			"type": "m.room.create",
+			"state_key": "",
+			"sender": "not_a_user_id",
+			"room_id": "!r1:a",
+			"event_id": "$e5:a",
+			"content": {"creator": "@u1:a"},
+			"unsigned": {
+				"not_allowed": "Sender is not a valid user ID"
+			}
+		}, {
+			"type": "m.room.create",
+			"state_key": "",
+			"sender": "@u1:a",
+			"room_id": "not_a_room_id",
+			"event_id": "$e6:a",
+			"content": {"creator": "@u1:a"},
+			"unsigned": {
+				"not_allowed": "Room is not a valid room ID"
+			}
+		}]
+	}`)
+}
+
+func TestAllowedWithNoPowerLevels(t *testing.T) {
+	testEventAllowed(t, `{
+		"auth_events": {
+			"create": {
+				"type": "m.room.create",
+				"sender": "@u1:a",
+				"room_id": "!r1:a",
+				"event_id": "$e1:a",
+				"content": {"creator": "@u1:a"}
+			},
+			"member": {
+				"@u1:a": {
+					"type": "m.room.member",
+					"sender": "@u1:a",
+					"room_id": "!r1:a",
+					"state_key": "@u1:a",
+					"event_id": "$e2:a",
+					"content": {"membership": "join"}
+				}
+			}
+		},
+		"allowed": [{
+			"type": "m.room.message",
+			"sender": "@u1:a",
+			"room_id": "!r1:a",
+			"event_id": "$e3:a",
+			"content": {"body": "Test"}
+		}],
+		"not_allowed": [{
+			"type": "m.room.message",
+			"sender": "@u2:a",
+			"room_id": "!r1:a",
+			"event_id": "$e4:a",
+			"content": {"body": "Test"},
+			"unsigned": {
+				"not_allowed": "Sender is not in room"
+			}
+		}]
+	}`)
+}
+
+func TestAllowedNoFederation(t *testing.T) {
+	testEventAllowed(t, `{
+		"auth_events": {
+			"create": {
+				"type": "m.room.create",
+				"sender": "@u1:a",
+				"room_id": "!r1:a",
+				"event_id": "$e1:a",
+				"content": {
+					"creator": "@u1:a",
+					"m.federate": false
+				}
+			},
+			"member": {
+				"@u1:a": {
+					"type": "m.room.member",
+					"sender": "@u1:a",
+					"room_id": "!r1:a",
+					"state_key": "@u1:a",
+					"event_id": "$e2:a",
+					"content": {"membership": "join"}
+				}
+			}
+		},
+		"allowed": [{
+			"type": "m.room.message",
+			"sender": "@u1:a",
+			"room_id": "!r1:a",
+			"event_id": "$e3:a",
+			"content": {"body": "Test"}
+		}],
+		"not_allowed": [{
+			"type": "m.room.message",
+			"sender": "@u2:b",
```

Comment:
I'd prefer if you actually just varied the server rather than the local part as well here, so `@u1:b` because it's the domain part we care about.